pvlib/ivtools.py

```diff
@@ -0,0 +1,215 @@
+    '''
+
+    try:
+        from PySAM import PySSC
+    except ImportError as e:
+        raise(e)
+
+    datadict = {'tech_model': '6parsolve', 'financial_model': 'none',
+                'celltype': celltype, 'Vmp': v_mp,
+                'Imp': i_mp, 'Voc': v_oc, 'Isc': i_sc, 'alpha_isc': alpha_sc,
+                'beta_voc': beta_voc, 'gamma_pmp': gamma_pmp,
+                'Nser': cells_in_series, 'Tref': temp_ref}
+
+    result = PySSC.ssc_sim_from_dict(datadict)
+    a_ref = result['a']
+    I_L_ref = result['Il']
+    I_o_ref = result['Io']
+    R_s = result['Rs']
+    R_sh_ref = result['Rsh']
+    Adjust = result['Adj']
+
+    return I_L_ref, I_o_ref, R_sh_ref, R_s, a_ref, Adjust
+
+
+def fit_sde_sandia(v, i, v_oc, i_sc, v_mp, i_mp, vlim=0.2, ilim=0.1):
+    """ Fits the single diode equation to an IV curve.
+
+    If fitting fails, returns NaN in each parameter.
+
+    Parameters
+    ----------
+    v : numeric
+        Voltage at each point on the IV curve, from 0 to v_oc
+
+    i : numeric
+        Current at each point on the IV curve, from i_sc to 0
+
+    v_oc : float
+        Open circuit voltage
+
+    i_sc : float
+        Short circuit current
+
+    v_mp : float
+        Voltage at maximum power point
+
+    i_mp : float
+        Current at maximum power point
+
+    vlim : float, default 0.2
+        defines linear portion of IV curve i.e. V <= vlim * v_oc
+
+    ilim : float, default 0.1
+        defines exponential portion of IV curve i.e. I > ilim * i_sc
+
+    Returns
+    -------
+    tuple of the following elements:
+
+        photocurrent : float
+            photocurrent, A
+
+        saturation_current : float
+            dark (saturation) current, A
+
+        resistance_shunt : float
+            shunt (parallel) resistance, ohm
+
+        resistance_series : float
+            series resistance, ohm
+
+        nNsVth : float
+            product of diode (ideality) factor n (unitless) x number of
+            cells in series Ns (unitless) x cell thermal voltage Vth (V), V
+
+    References
+    ----------
+    [1] C. B. Jones, C. W. Hansen, Single Diode Parameter Extraction from
+    In-Field Photovoltaic I-V Curves on a Single Board Computer, 46th IEEE
+    Photovoltaic Specialist Conference, Chicago, IL, 2019
+    """
+    # Find intercept and slope of linear portion of IV curve.
+    # Start with V < vlim * v_oc, extend by adding points until slope is
+    # acceptable
+    beta = [np.nan for i in range(5)]
+    # get index of largest voltage less than/equal to limit
+    idx = _max_index(v, vlim * v_oc)
+    while np.isnan(beta[1]) and (idx <= len(v)):
+        try:
+            coef = np.polyfit(v[:idx], i[:idx], deg=1)
+            if coef[0] < 0:
+                # intercept term
+                beta[0] = coef[1].item()
+                # sign change of slope to get positive parameter value
+                beta[1] = -coef[0].item()
+        except Exception as e:
```

Comment:
So far as I can tell, these `try; except Exception as e; raise e` blocks don't serve any purpose.

Reply:
You're right.  Either provide a message that helps a pvlib user understand why the np.polyfit error comes about, or delete and let the np.polyfit error stand alone.

Reply:
`numpy.polyfit` raises `ValueError` or `TypeError` when inputs are wrong shape or empty.  I think that's sufficient.  `numpy.polyfit` raises `RankWarning` when the fit is suspect.